lib/vagrant/action/builder.rb

```diff
@@ -132,42 +133,153 @@ def index(object)
       # Converts the builder stack to a runnable action sequence.
       #
       # @param [Hash] env The action environment hash
-      # @return [Object] A callable object
+      # @return [Warden] A callable object
       def to_app(env)
-        app_stack = nil
+        # Start with a duplicate of ourself which can
+        # be modified
+        builder = self.dup
 
-        # If we have action hooks, then we apply them
-        if env[:action_hooks]
-          builder = self.dup
+        # Apply all dynamic modifications of the stack. This
+        # will generate dynamic hooks for all actions within
+        # the stack, load any triggers for action classes, and
+        # apply them to the builder's stack
+        builder.apply_dynamic_updates(env)
 
-          # These are the options to pass into hook application.
-          options = {}
+        # Now that the stack is fully expanded, apply any
+        # action hooks that may be defined so they are on
+        # the outermost locations of the stack
+        builder.apply_action_name(env)
 
-          # If we already ran through once and did append/prepends,
-          # then don't do it again.
-          if env[:action_hooks_already_ran]
-            options[:no_prepend_or_append] = true
+        # Wrap the middleware stack with the Warden to provide a consistent
+        # and predictable behavior upon exceptions.
+        Warden.new(builder.stack.dup, env)
+      end
+
+      # Find any action hooks or triggers which have been defined
+      # for items within the stack. Update the stack with any
+      # hooks or triggers found.
+      #
+      # @param [Hash] env Call environment
+      # @return [Builder] self
+      def apply_dynamic_updates(env)
```

Comment:
Do we need tests for these new methods? #apply_dynamic_updates and #apply_action_name

Reply:
Yes, yes we do.